server/controllers/stock/shipment.js

```diff
@@ -0,0 +1,214 @@
+const db = require('../../lib/db');
+const FilterParser = require('../../lib/filter');
+const { uuid } = require('../../lib/util');
+
+const SHIPMENT_PARTIAL = 2;
+const SHIPMENT_COMPLETE = 3;
+const SHIPMENT_IN_TRANSIT = 4;
+const SHIPMENT_IN_TRANSIT_OR_PARTIAL = [SHIPMENT_IN_TRANSIT, SHIPMENT_PARTIAL];
+
+exports.writeStockExitShipment = (
+  projectId,
+  from,
+  to,
+  document,
+  parameters,
+  transaction,
+) => {
+  const SHIPMENT_UUID = db.bid(uuid());
+  const SHIPMENT_LABEL = 'Depot Exit Shipment';
+  const TRANSIT_SHIPPER = 1;
+  const shipment = {
+    uuid : SHIPMENT_UUID,
+    name : SHIPMENT_LABEL,
+    project_id : projectId,
+    description : parameters.description,
+    shipper_id : TRANSIT_SHIPPER,
+    origin_depot_uuid : from,
+    current_depot_uuid : from,
+    destination_depot_uuid : to,
+    anticipated_delivery_date : document.date,
+    date_sent : document.date,
+    status_id : SHIPMENT_IN_TRANSIT,
+    created_by : document.user,
+    document_uuid : db.bid(document.uuid),
+  };
+  transaction.addQuery('INSERT INTO shipment SET ?', [shipment]);
+
+  parameters.lots.forEach((lot) => {
+    const shipmentItem = {
+      uuid : db.bid(uuid()),
+      shipment_uuid : SHIPMENT_UUID,
+      lot_uuid : db.bid(lot.uuid),
+      date_sent : document.date,
+      quantity_sent : lot.quantity,
+    };
+    transaction.addQuery('INSERT INTO shipment_item SET ?', [shipmentItem]);
+  });
+};
+
+exports.writeStockEntryShipment = async (
+  document,
+  parameters,
+  transaction,
+) => {
+  // update shipment items
+  parameters.lots.forEach(lot => {
+    const updateShipmentItem = `
```

Comment:
This `updateShipmentItem` doesn't change for  each lot.  You can move it outside the loop.


```js
const updateShipmentItem = `
        UPDATE shipment_item shi
        JOIN shipment sh ON sh.uuid = shi.shipment_uuid 
        SET
          shi.quantity_delivered = shi.quantity_delivered + ?,
          shi.date_delivered = ?,
          sh.date_delivered = ?
        WHERE sh.status_id IN (?) AND sh.document_uuid = ? AND shi.lot_uuid = ?
`;

parameters.lots.forEach(lot => {
  transaction.addQuery(updateShipmentItem, [ 
    lot.quantity || 0,
    document.date,
    document.date,
    SHIPMENT_IN_TRANSIT_OR_PARTIAL,
    db.bid(document.uuid),
    db.bid(lot.uuid),
  ]);
});
```

Reply:
RESOLVED 👍🏽